File: plox/src/interpreter/interpreter.py
# Project Imports
from src.ast.expr import (
    Expr,
    ExprVisitor,
    Assign,
    Binary,
    Call,
    Conditional,
    Grouping,
    Literal,
    Logical,
    Unary,
    Variable
)
from src.ast.stmt import (
    Stmt,
    StmtVisitor,
    Block,
    Break,
    Const,
    Continue,
    Echo,
    Expression,
    For,
    Function,
    If,
    Return,
    Var,
    While
)
from src.callable.lox_callable import LoxCallable
from src.callable.lox_function import LoxFunction
from src.interpreter.environment import Environment
from src.callable.natives import define_natives
from src.scanner.token import TokenType
from src.util.errors import LoxRuntimeError
from src.util.exceptions import (
    BreakException,
    ContinueException,
    ReturnException
)
from src.util.mode import RunMode


# Interpreter Class
class Interpreter(ExprVisitor, StmtVisitor):
    def __init__(self, err_manager, mode):
        self.err_manager = err_manager
        self.mode = mode
        self.globals = Environment()
        self.environment = self.globals
        self.locals = {}
        self.loop_depth = 0

        # Native Functions
        define_natives(self)

    def interpret(self, statements: list[Stmt]):
        try:
            for statement in statements:
                if self.mode == RunMode.REPL:
                    self.execute_repl_friendly(statement)
                elif self.mode == RunMode.FILE:
                    self.execute(statement)
        except LoxRuntimeError as e:
            self.err_manager.runtime_error(e)

    def execute_repl_friendly(self, stmt: Stmt):
        if isinstance(stmt, Expression):
            if not isinstance(stmt.expression, Assign):
                value = self.evaluate(stmt.expression)
                print(self.stringify(value))
            else:
                self.evaluate(stmt.expression)
        else:
            self.execute(stmt)

    def resolve(self, expr, depth):
        self.locals[expr] = depth

    def execute(self, stmt: Stmt):
        stmt.accept(self)

    def execute_block(self, statements: list[Stmt], environment):
        previous = self.environment

        try:
            self.environment = environment

            for statement in statements:
                self.execute(statement)
        finally:
            self.environment = previous

    def evaluate(self, expr: Expr):
        return expr.accept(self)

    def visit_block_stmt(self, stmt: Block):
        self.execute_block(stmt.statements, Environment(self.environment))

    def visit_break_stmt(self, stmt: Break):
        raise BreakException(stmt.keyword)

    def visit_const_stmt(self, stmt: Const):
        value = self.evaluate(stmt.initializer)
        self.environment.define_const(stmt.name.lexeme, value)

    def visit_continue_stmt(self, stmt: Continue):
        raise ContinueException(stmt.keyword)

    def visit_echo_stmt(self, stmt: Echo):
        value = self.evaluate(stmt.expression)
        print(self.stringify(value))

    def visit_expression_stmt(self, stmt: Expression):
        self.evaluate(stmt.expression)

    def visit_for_stmt(self, stmt: For):
        if stmt.initializer != None:
            self.evaluate(stmt.initializer)

        try:
            self.loop_depth += 1
            while self.is_truthy(self.evaluate(stmt.condition)):
                try:
                    self.evaluate(stmt.body)
                except BreakException:
                    return
                except ContinueException:
                    if isinstance(stmt.body, Block):
                        self.execute_block(
                            [stmt.body.statements[-1]],
                            Environment(self.environment)
                        )
        finally:
            self.loop_depth -= 1

    def visit_function_stmt(self, stmt: Function):
        function = LoxFunction(stmt, self.environment)
        self.environment.define_const(stmt.name.lexeme, function)

    def visit_if_stmt(self, stmt: If):
        if self.is_truthy(self.evaluate(stmt.condition)):
            self.execute(stmt.then_branch)

        elif stmt.else_branch != None:
            self.execute(stmt.else_branch)

    def visit_return_stmt(self, stmt: Return):
        value = None
        if stmt.value != None:
            value = self.evaluate(stmt.value)

        raise ReturnException(stmt.keyword.lexeme, value)

    def visit_var_stmt(self, stmt: Var):
        value = None
        if stmt.initializer != None:
            value = self.evaluate(stmt.initializer)

        self.environment.define(stmt.name.lexeme, value)

    def visit_while_stmt(self, stmt: While):
        while self.is_truthy(self.evaluate(stmt.condition)):
            self.execute(stmt.body)

    def visit_assign_expr(self, expr: Assign):
        value = self.evaluate(expr.value)

        match expr.operator.type:
            case TokenType.EQ:
                distance = self.locals.get(expr)
                if distance != None:
                    self.environment.assign_at(distance, expr.name, value)
                else:
                    self.environment.assign(expr.name, value)

                return value

            case TokenType.MINUSEQ:
                if not isinstance(value, float):
                    raise LoxRuntimeError(expr.operator, "Cannot use augmented assignment on non-number values.")

                initial = self.environment.get(expr.name)
                initial -= value

                distance = self.locals.get(expr)
                if distance != None:
                    self.environment.assign_at(distance, expr.name, initial)
                else:
                    self.environment.assign(expr.name, initial)

                return value

            case TokenType.MODEQ:
                if not isinstance(value, float):
                    raise LoxRuntimeError(expr.operator, "Cannot use augmented assignment on non-number values.")

                if value == 0:
                    raise LoxRuntimeError(expr.operator, "Cannot divide by Zero.")

                initial = self.environment.get(expr.name)
                initial %= value

                distance = self.locals.get(expr)
                if distance != None:
                    self.environment.assign_at(distance, expr.name, initial)
                else:
                    self.environment.assign(expr.name, initial)

                return value

            case TokenType.PLUSEQ:
                if not isinstance(value, float):
                    raise LoxRuntimeError(expr.operator, "Cannot use augmented assignment on non-number values.")

                initial = self.environment.get(expr.name)
                initial += value

                distance = self.locals.get(expr)
                if distance != None:
                    self.environment.assign_at(distance, expr.name, initial)
                else:
                    self.environment.assign(expr.name, initial)

                return value

            case TokenType.SLASHEQ:
                if not isinstance(value, float):
                    raise LoxRuntimeError(expr.operator, "Cannot use augmented assignment on non-number values.")

                if value == 0:
                    raise LoxRuntimeError(expr.operator, "Cannot divide by Zero.")

                initial = self.environment.get(expr.name)
                initial /= value

                distance = self.locals.get(expr)
                if distance != None:
                    self.environment.assign_at(distance, expr.name, initial)
                else:
                    self.environment.assign(expr.name, initial)

                return value

            case TokenType.STAREQ:
                if not isinstance(value, float):
                    raise LoxRuntimeError(expr.operator, "Cannot use augmented assignment on non-number values.")

                initial = self.environment.get(expr.name)
                initial *= value

                distance = self.locals.get(expr)
                if distance != None:
                    self.environment.assign_at(distance, expr.name, initial)
                else:
                    self.environment.assign(expr.name, initial)

                return value

    def visit_binary_expr(self, expr: Binary):
        left = self.evaluate(expr.left)
        right = self.evaluate(expr.right)

        match expr.operator.type:
            case TokenType.MINUS:
                if self.check_operands(expr.operator, left, right) == True:
                    return left - right

            case TokenType.MODULUS:
                if self.check_operands(expr.operator, left, right) == True:
                    if right == 0:
                        raise LoxRuntimeError(expr.operator, "Cannot divide by Zero.")

                    return left % right

            case TokenType.PLUS:
                if isinstance(left, float) and isinstance(right, float):
                    return left + right

                elif isinstance(left, str) and isinstance(right, str):
                    return left + right

                elif isinstance(left, float) and isinstance(right, str):
                    return self.stringify(left) + right

                elif isinstance(left, str) and isinstance(right, float):
                    return left + self.stringify(right)

                raise LoxRuntimeError(expr.operator, "Operands must be numbers or strings. Combining the two is allowed.")
            case TokenType.POWER:
                if self.check_operands(expr.operator, left, right) == True:
                    return left ** right

            case TokenType.SLASH:
                if self.check_operands(expr.operator, left, right) == True:
                    if right == 0:
                        raise LoxRuntimeError(expr.operator, "Cannot divide by Zero.")

                    return left / right

            case TokenType.STAR:
                if self.check_operands(expr.operator, left, right) == True:
                    return left * right

            case TokenType.GT:
                if self.check_operands(expr.operator, left, right) == True:
                    return left > right

            case TokenType.GTEQ:
                if self.check_operands(expr.operator, left, right) == True:
                    return left >= right

            case TokenType.LT:
                if self.check_operands(expr.operator, left, right) == True:
                    return left < right

            case TokenType.LTEQ:
                if self.check_operands(expr.operator, left, right) == True:
                    return left <= right

            case TokenType.BANGEQ:
                return not self.is_equal(left, right)

            case TokenType.EQEQ:
                return self.is_equal(left, right)

    def visit_call_expr(self, expr: Call):
        callee = self.evaluate(expr.callee)

        arguments = []
        for argument in expr.arguments:
            arguments.append(self.evaluate(argument))

        if not isinstance(callee, LoxCallable):
            raise LoxRuntimeError(
                expr.paren,
                "Only classes, functions or methods can be called."
            )

        function = callee
        if len(arguments) != function.arity():
            raise LoxRuntimeError(
                expr.paren,
                f"Expected {function.arity()} arguments " +
                f"but got {len(arguments)} instead."
            )
        return function.call(self, arguments)

    def visit_conditional_expr(self, expr: Conditional):
        if self.is_truthy(self.evaluate(expr.condition)):
            return self.evaluate(expr.then_branch)

        return self.evaluate(expr.else_branch)

    def visit_grouping_expr(self, expr: Grouping):
        return self.evaluate(expr.expression)

    def visit_literal_expr(self, expr: Literal):
        return expr.value

    def visit_logical_expr(self, expr: Logical):
        left = self.evaluate(expr.left)

        if expr.operator.type == TokenType.OR:
            if self.is_truthy(left):
                return left
        else:
            if not self.is_truthy(left):
                return left

        return self.evaluate(expr.right)

    def visit_unary_expr(self, expr: Unary):
        right = self.evaluate(expr.right)

        match expr.operator.type:
            case TokenType.BANG:
                return not self.is_truthy(right)

            case TokenType.MINUS:
                if self.check_operand(expr.operator, right) == True:
                    return -right

    def visit_variable_expr(self, expr: Variable):
        # return self.environment.get(expr.name)
        return self.look_up_variable(expr.name, expr)

    def look_up_variable(self, name, expr):
        distance = self.locals.get(expr)
        if distance != None:
            return self.environment.get_at(distance, name.lexeme)
        else:
            return self.globals.get(name)

    def is_truthy(self, obj):
        if obj == None:
            return False

        if isinstance(obj, bool):
            return obj

        return True

    def is_equal(self, a, b):
        if (a == None) and (b == None):
            return True

        if a == None:
            return False

        return a == b

    def check_operand(self, operator, operand):
        if isinstance(operand, float):
            return True

        raise LoxRuntimeError(operator, "Operand must be a number.")

    def check_operands(self, operator, left, right):
        if isinstance(left, float) and isinstance(right, float):
            return True

        raise LoxRuntimeError(operator, "Operands must be numbers.")

    def stringify(self, obj):
        if obj == None:
            return "null"

        if isinstance(obj, float):
            text = str(obj)
            if text.endswith(".0"):
                text = text[0 : len(text) - 2]

            return text

        return str(obj)

File: jlox/src/lox/Interpreter.java
package lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Interpreter implements Expr.Visitor<Object>, Stmt.Visitor<Void> {
    final Environment globals = new Environment();
    private Environment environment = globals;
    private final Map<Expr, Integer> locals = new HashMap<>();

    Interpreter() {
        globals.define("clock", new LoxCallable() {
            @Override
            public int
            arity() { return 0; }

            @Override
            public Object
            call(Interpreter interpreter, List<Object> arguments)
            {
                return (double)System.currentTimeMillis() / 1000.0;
            }

            @Override
            public String
            toString() { return "<Native Fn - clock>"; }
        });
    }

    void
    interpret(List<Stmt> statements)
    {
        try {
            for (Stmt statement : statements) {
                execute(statement);
            }
        } catch (RuntimeError error) {
            Lox.runtimeError(error);
        }
    }

    void
    resolve(Expr expr, int depth)
    {
        locals.put(expr, depth);
    }

    private void
    execute(Stmt stmt)
    {
        stmt.accept(this);
    }

    void
    executeBlock(List<Stmt> statements, Environment environment)
    {
        Environment previous = this.environment;

        try {
            this.environment = environment;

            for (Stmt statement : statements) {
                execute(statement);
            }
        } finally {
            this.environment = previous;
        }
    }

    private Object
    evaluate(Expr expr)
    {
        return expr.accept(this);
    }

    @Override
    public Void
    visitBlockStmt(Stmt.Block stmt)
    {
        executeBlock(stmt.statements, new Environment(environment));
        return null;
    }

    @Override
    public Void
    visitClassStmt(Stmt.Class stmt)
    {
        Object superclass = null;
        if (stmt.superclass != null) {
            superclass = evaluate(stmt.superclass);
            if (!(superclass instanceof LoxClass)) {
                throw new RuntimeError(stmt.superclass.name, "Superclass must be a class.");
            }
        }

        environment.define(stmt.name.lexeme, null);

        if (stmt.superclass != null) {
            environment = new Environment(environment);
            environment.define("super", superclass);
        }

        Map<String, LoxFunction> methods = new HashMap<>();
        for (Stmt.Function method : stmt.methods) {
            LoxFunction function = new LoxFunction(
                    method,
                    environment,
                    method.name.lexeme.equals("init")
            );

            methods.put(method.name.lexeme, function);
        }

        LoxClass klass = new LoxClass(stmt.name.lexeme, (LoxClass)superclass, methods);

        if (superclass != null) {
            environment = environment.enclosing;
        }

        environment.assign(stmt.name, klass);
        return null;
    }

    @Override
    public Void
    visitExpressionStmt(Stmt.Expression stmt)
    {
        evaluate(stmt.expression);
        return null;
    }

    @Override
    public Void
    visitFunctionStmt(Stmt.Function stmt)
    {
        LoxFunction function = new LoxFunction(stmt, environment, false);
        environment.define(stmt.name.lexeme, function);
        return null;
    }

    @Override
    public Void
    visitIfStmt(Stmt.If stmt)
    {
        if (isTruthy(evaluate(stmt.condition))) {
            execute(stmt.thenBranch);
        } else if (stmt.elseBranch != null) {
            execute(stmt.elseBranch);
        }

        return null;
    }

    @Override
    public Void
    visitPrintStmt(Stmt.Print stmt)
    {
        Object value = evaluate(stmt.expression);
        System.out.println(stringify(value));
        return null;
    }

    @Override
    public Void
    visitReturnStmt(Stmt.Return stmt)
    {
        Object value = null;
        if (stmt.value != null) value = evaluate(stmt.value);

        throw new Return(value);
    }

    @Override
    public Void visitVarStmt(Stmt.Var stmt) {
        Object value = null;
        if (stmt.initializer != null) {
            value = evaluate(stmt.initializer);
        }

        environment.define(stmt.name.lexeme, value);
        return null;
    }

    @Override
    public Void
    visitWhileStmt(Stmt.While stmt)
    {
        while (isTruthy(evaluate(stmt.condition))) {
            execute(stmt.body);
        }

        return null;
    }

    @Override
    public Object
    visitAssignExpr(Expr.Assign expr)
    {
        Object value = evaluate(expr.value);

        Integer distance = locals.get(expr);
        if (distance != null) {
            environment.assignAt(distance, expr.name, value);
        } else {
            environment.assign(expr.name, value);
        }

        return value;
    }

    @Override
    public Object
    visitBinaryExpr(Expr.Binary expr)
    {
        Object left = evaluate(expr.left);
        Object right = evaluate(expr.right);

        switch (expr.operator.type) {
            case MINUS -> {
                checkOperands(expr.operator, left, right);
                return (double) left - (double) right;
            }
            case SLASH -> {
                checkOperands(expr.operator, left, right);

                if ((double)right == 0) {
                    throw new RuntimeError(expr.operator, "Cannot divide by Zero.");
                }

                return (double) left / (double) right;
            }
            case STAR -> {
                checkOperands(expr.operator, left, right);
                return (double) left * (double) right;
            }
            case PLUS -> {
                if (left instanceof Double && right instanceof Double) {
                    return (double)left + (double)right;
                }

                if (left instanceof String && right instanceof String) {
                    return left + (String)right;
                }

                if (left instanceof Double && right instanceof String) {
                    return stringify(left) + right;
                }

                if (left instanceof String && right instanceof Double) {
                    return left + stringify(right);
                }

                throw new RuntimeError(expr.operator, "Operands must be numbers or strings, not mixed.");
            }
            case GT -> {
                checkOperands(expr.operator, left, right);
                return (double) left > (double) right;
            }
            case GTEQ -> {
                checkOperands(expr.operator, left, right);
                return (double) left >= (double) right;
            }
            case LT -> {
                checkOperands(expr.operator, left, right);
                return (double) left < (double) right;
            }
            case LTEQ -> {
                checkOperands(expr.operator, left, right);
                return (double) left <= (double) right;
            }
            case BANGEQ -> {
                return !isEqual(left, right);
            }
            case EQEQ -> {
                return isEqual(left, right);
            }
        }

        return null;
    }

    @Override
    public Object
    visitCallExpr(Expr.Call expr)
    {
        Object callee = evaluate(expr.callee);

        List<Object> arguments = new ArrayList<>();
        for (Expr argument : expr.arguments) {
            arguments.add(evaluate(argument));
        }

        if (!(callee instanceof LoxCallable function)) {
            throw new RuntimeError(expr.paren, "Can only call classes, functions or methods.");
        }

        if (arguments.size() != function.arity()) {
            throw new RuntimeError(expr.paren,
                    "Expected " + function.arity() + " arguments" +
                    " but got " + arguments.size() + " instead.");
        }

        return function.call(this, arguments);
    }

    @Override
    public Object
    visitGetExpr(Expr.Get expr)
    {
        Object object = evaluate(expr.object);
        if (object instanceof LoxInstance) {
            return ((LoxInstance)object).get(expr.name);
        }

        throw new RuntimeError(expr.name, "Only instances of an object have properties.");
    }

    @Override
    public Object
    visitGroupingExpr(Expr.Grouping expr)
    {
        return evaluate(expr.expression);
    }

    @Override
    public Object
    visitLiteralExpr(Expr.Literal expr)
    {
        return expr.value;
    }

    @Override
    public Object
    visitLogicalExpr(Expr.Logical expr)
    {
        Object left = evaluate(expr.left);

        if (expr.operator.type == TokenType.OR) {
            if (isTruthy(left)) return left;
        } else {
            if (!isTruthy(left)) return left;
        }

        return evaluate(expr.right);
    }

    @Override
    public Object
    visitSetExpr(Expr.Set expr)
    {
        Object object = evaluate(expr.object);

        if (!(object instanceof LoxInstance)) {
            throw new RuntimeError(expr.name, "only instances have fields.");
        }

        Object value = evaluate(expr.value);
        ((LoxInstance)object).set(expr.name, value);

        return value;
    }

    @Override
    public Object
    visitSuperExpr(Expr.Super expr)
    {
        int distance = locals.get(expr);

        LoxClass superclass = (LoxClass)environment.getAt(distance, "super");
        LoxInstance object = (LoxInstance)environment.getAt(distance - 1, "this");
        LoxFunction method = superclass.findMethod(expr.method.lexeme);

        if (method == null) {
            throw new RuntimeError(expr.method, "Undefined property '" + expr.method.lexeme + "'.");
        }

        return method.bind(object);
    }

    @Override
    public Object
    visitThisExpr(Expr.This expr)
    {
        return lookUpVariable(expr.keyword, expr);
    }

    @Override
    public Object
    visitUnaryExpr(Expr.Unary expr)
    {
        Object right = evaluate(expr.right);

        return switch (expr.operator.type) {
            case BANG -> !isTruthy(right);
            case MINUS -> {
                checkOperand(expr.operator, right);
                yield -(double) right;
            }
            default -> null;
        };
    }

    @Override
    public Object
    visitVariableExpr(Expr.Variable expr)
    {
        return lookUpVariable(expr.name, expr);
    }

    private Object
    lookUpVariable(Token name, Expr expr)
    {
        Integer distance = locals.get(expr);

        if (distance != null) {
            return environment.getAt(distance, name.lexeme);
        } else {
            return globals.get(name);
        }
    }

    private boolean
    isTruthy(Object object)
    {
        if (object == null) return false;
        if (object instanceof Boolean) return (boolean)object;
        return true;
    }

    private boolean
    isEqual(Object a, Object b)
    {
        if (a == null && b == null) return true;
        if (a == null) return false;

        return a.equals(b);
    }

    private void
    checkOperand(Token operator, Object operand)
    {
        if (operand instanceof Double) return;
        throw new RuntimeError(operator, "Operand must be a number.");
    }

    private void
    checkOperands(Token operator, Object left, Object right)
    {
        if (left instanceof Double && right instanceof Double) return;
        throw new RuntimeError(operator, "Operands must be numbers.");
    }

    private String
    stringify(Object object)
    {
        if (object == null) return "nil";

        if (object instanceof Double) {
            String text = object.toString();
            if (text.endsWith(".0")) {
                text = text.substring(0, text.length() - 2);
            }

            return text;
        }

        return object.toString();
    }
}
